src/lib/Serialize.hs

```diff
@@ -224,14 +232,68 @@ liftExceptIO :: Except a -> IO a
 liftExceptIO (Left e ) = throwIO e
 liftExceptIO (Right x) = return x
 
-flattenType :: Type -> [(BaseType, [Int])]
-flattenType (FixedIntRange _ _) = [(IntType, [])]
-flattenType (TabTy (FixedIntRange low high) a) =
-    [(b, (high - low):shape) | (b, shape) <- flattenType a]
--- temporary hack. TODO: fix
-flattenType (TabTy _ a) = [(b, 0:shape) | (b, shape) <- flattenType a]
+indexSetSize :: Type -> Int
+indexSetSize t = fromMaybe fail' $ indexSetConcreteSize t
+  where
+    fail' = (error $ "Can't determine dimension size: " ++ pprint t)
+    indexSetConcreteSize :: Type -> Maybe Int
+    indexSetConcreteSize ty = case ty of
+      TC (IntRange low high) -> do
+        l <- loadBound low
+        h <- loadBound high
+        Just $ h - l
+      TC (IndexRange n low high) -> do
+        low' <- case low of
+          InclusiveLim x -> loadBound x
+          ExclusiveLim x -> (+1) <$> loadBound x
+          Unlimited      -> Just 0
+        high' <- case high of
+          InclusiveLim x -> (+1) <$> loadBound x
+          ExclusiveLim x -> loadBound x
+          Unlimited      -> indexSetConcreteSize n
+        Just $ high' - low'
+      BoolTy  -> Just 2
+      RecTy r -> liftM product $ mapM indexSetConcreteSize $ toList r
+      SumTy l r -> (+) <$> indexSetConcreteSize l <*> indexSetConcreteSize r
+      _ -> Nothing
+      where
+        loadBound :: Atom -> Maybe Int
+        loadBound (IntVal n) = Just n
+        loadBound (Con (AGet (Con (ArrayLit arr)))) = do
+          (IntLit n) <- scalarFromArray arr
+          Just n
+        loadBound _ = Nothing
+
+-- Similar to flattenType, except it only returns the missing arrays
+flattenVal :: Val -> [(Var, ArrayType)]
+flattenVal = flattenVal' 1
+
+flattenVal' :: Int -> Val -> [(Var, ArrayType)]
+-- TODO: Make the JAX backend return atoms with only ArrayTy
```

Comment:
TODO: JAX backend currently doesn't use `ArrayTy`, because I haven't implemented the type recovery based on the AFor statements. This is left as a future cleanup.